### PR TITLE
RavenDB-11552 Deferred commands do not affect session state

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 1]
```diff
@@ -77,21 +80,154 @@ CommandType GetCommandType(BlittableJsonReaderObject batchResult)
...
+                ThrowMissingField(CommandType.PATCH, nameof(PatchStatus));
+
+            if (Enum.TryParse(statusAsString, ignoreCase: true, out PatchStatus status) == false)
+                ThrowMissingField(CommandType.PATCH, nameof(PatchStatus));
+
+            switch (status)
+            {
+                case PatchStatus.Created:
+                case PatchStatus.Patched:
+                    HandleDeleteInternal(batchResult, CommandType.PATCH); // deleting because it is impossible to know what was in the patch
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: @ayende @maximburyak not sure what to do here. What do you think?

**Comment by ppekrol** [comment]
@maximburyak Can you review?

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 2]
```diff
@@ -77,21 +80,154 @@ CommandType GetCommandType(BlittableJsonReaderObject batchResult)
...
+
+            var attachments = documentInfo.Metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachmentsJson)
+                ? new DynamicJsonArray(attachmentsJson)
+                : new DynamicJsonArray();
+
+            documentInfo.Metadata.Modifications[Constants.Documents.Metadata.Attachments] = attachments;
+
+            attachments.Add(new DynamicJsonValue
+            {
+                [nameof(AttachmentDetails.ChangeVector)] = GetStringField(batchResult, type, nameof(AttachmentDetails.ChangeVector)),
```
Comment (maximburyak) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: We can avoid the string allocation here, we could just use the LazyStringValue

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 3]
```diff
@@ -77,21 +80,154 @@ CommandType GetCommandType(BlittableJsonReaderObject batchResult)
...
-            // DeferredCommands
-            for (var i = _sessionCommandsCount; i < result.Results.Length; i++)
+            if (_session.DocumentsById.TryGetValue(id, out var documentInfo) == false)
+                return;
+
+            if (documentInfo.Metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachmentsJson) == false || attachmentsJson == null ||
+                attachmentsJson.Length == 0)
+                return;
+
+            var name = GetStringField(batchResult, type, attachmentNameFieldName);
```
Comment (maximburyak) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: We could avoid string materialization here, by directly comparing the attachmentNameFieldName with attachmentName, as they are both LazyStringValues

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 4]
```diff
@@ -154,6 +295,27 @@ private void UpdateCounterValuesInCache(BlittableJsonReaderArray counters, strin
             }
         }
 
+        private static string GetStringField(BlittableJsonReaderObject json, CommandType type, string fieldName)
```
Comment (maximburyak) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: I recommend to see where we can avoid using that function

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs (hunk 3 above)
Comment (maximburyak) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: We could have a direct comparison between this and the name from Attachment in line 150

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs (hunk 2 above)
Comment (maximburyak) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: We could use LazyStringValue directrly here

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs (hunk 1 above)
Comment (maximburyak) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: It makes sense, don't think we should do anything else here

**Comment by ppekrol** [comment]
@maximburyak nice catch!

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 5]
```diff
@@ -20,13 +22,17 @@ public BatchOperation(InMemoryDocumentSessionOperations session)
 
         private List<object> _entities;
         private int _sessionCommandsCount;
+        private int _allCommandsCount;
 
         public BatchCommand CreateRequest()
         {
             var result = _session.PrepareForSaveChanges();
             _sessionCommandsCount = result.SessionCommands.Count;
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs:31]: Why is this needed now?

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 6]
```diff
@@ -64,48 +70,192 @@ CommandType GetCommandType(BlittableJsonReaderObject batchResult)
                         $"Cluster transaction was send to a node that is not supporting it. So it was executed ONLY on the requested node on {_session.RequestExecutor.Url}.");
             }
 
-            for (var i = 0; i < _sessionCommandsCount; i++)
+            for (var i = 0; i < _allCommandsCount; i++)
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: Better to have two `for` statements, one from `0 .. _sessionCommandsCount` and one from `_sessionCommandsCount` to `_allCommandsCount`

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: This will mean that loading the document again will _return null_, no?
In this case, probably should evict it.

Note that I'm pretty sure that we already check that the document isn't already in the session / modified.

Comment (ppekrol) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: This will remove it from the session (evict). So next 'Load' will issue a new request to the server to get it.

Downside of this is that following code fill not work:
```csharp
using (session)
{
    var x = load
    patch(x)
    saveChanges

    x.Name = 'Name' // not tracked anymore
    saveChanges // nothing will happen
}
```

Comment (ayende) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: _should_ such code work?
That is a very strange thing to do, I think.
I also think we check if there are modifications to a patched entity.
I'm fine with keeping that check across `SaveChanges` calls.

Maybe just mark this as patched, meaning that unless user explicitly evict, we can't user the existing instance?

Comment (ppekrol) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: What if... just before sending the DeferredPatchCommand... we will check if entity exists on the session. And if it does, then issue this command with additional parameter to return the document after patch in the 'replay'. We can then update the session properly?

Comment (ayende) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: Sounds good to me. Yes.

Comment (ppekrol) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: Done

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs (hunk 6 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: Done

**Review by ayende — APPROVED** [review]
